components/api_request_helper/api_request_helper.cc

```diff
@@ -417,24 +453,36 @@ void APIRequestHelper::URLLoaderHandler::ParseSSE(
   static constexpr char kDataPrefix[] = "data: {";
 
   for (const auto& data : stream_data) {
+    DVLOG(3) << "Received chunk: " << data;
     if (!base::StartsWith(data, kDataPrefix)) {
+      VLOG(1) << "Data did not start with SSE prefix";
       continue;
     }
 
     std::string json = data.substr(strlen(kDataPrefix) - 1);
-
     auto on_json_parsed =
         [](base::WeakPtr<APIRequestHelper::URLLoaderHandler> handler,
            data_decoder::DataDecoder::ValueOrError result) {
+          VLOG(1) << "on_json_parsed";
           if (!handler) {
             return;
           }
+          handler->current_decoding_operation_count_--;
           DCHECK(handler->data_received_callback_);
+          VLOG(1) << "calling received";
           handler->data_received_callback_.Run(std::move(result));
+          if (!handler->data_chunk_parsed_callback_.is_null()) {
+            handler->data_chunk_parsed_callback_.Run();
```

Comment:
Could we check `current_decoding_operation_count_` and call `Cancel()` directly, not via `data_chunk_parsed_callback_` callback?
The  `url_loader_` status (was `OnCompleted` called or not) could be check using a method like `CompletionStatus()`
https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/cpp/simple_url_loader.h;bpv=1;bpt=1;l=369

That will make the code more clear/

Reply:
Here's a version of this. I called it `TryDelete` to note how it's about deletion after completion and not for Cancel which is perfectly ok to happen mid-operation.

Reply:
Here's a version of this. I called it `TryDelete` to note how it's about deletion after completion and not for Cancel which is perfectly ok to happen mid-operation. https://github.com/brave/brave-core/pull/18855/commits/db4c7816e6059d9272f007cbd5b732e11ea9a8b5

Reply:
Actually this isn't working well @atuchin-m as it throws a `CHECK` if we ask for completion status before the request is complete. The method comments state "May only be called once the loader has informed the caller of completion." So maybe I have to set a flag.

Reply:
Updated https://github.com/brave/brave-core/pull/18855/commits/9d3902e1201c5c94a2b76abc3a072d2ec89edd09 which now uses a flag